resources/net.go

```diff
@@ -0,0 +1,665 @@
+	// messages left in the channel
+	ifaceChan := make(chan *dbus.Signal, 10) // iface event channel
+	defer close(ifaceChan)
+
+	bus.Signal(ifaceChan)
+	defer bus.RemoveSignal(ifaceChan)
+
+	// build the path to the configuration file
+	obj.unitFile = NetworkdUnitFileDir + IfacePrefix + obj.GetName() + NetworkdUnitFileExt
+
+	// watch the configuration file
+	recWatcher, err := recwatch.NewRecWatcher(obj.unitFile, false)
+	if err != nil {
+		return err
+	}
+	defer recWatcher.Close()
+
+	// notify engine that we're running
+	if err := obj.Running(); err != nil {
+		return err // bubble up a NACK...
+	}
+
+	var send = false
+	var exit *error
+
+	for {
+		select {
+		case event := <-ifaceChan:
+			if len(event.Body) < 2 {
+				return fmt.Errorf("unexpected event body length")
+			}
+			// parse the event and save the operational state to the struct
+			bodyMap, ok := event.Body[1].(map[string]dbus.Variant)
+			if !ok {
+				return fmt.Errorf("error casting body map")
+			}
+			if val, ok := bodyMap["OperationalState"]; ok {
+				obj.oState = val
+			}
+			if obj.debug { // don't access event.Body if event.Error isn't nil
+				log.Printf("%s: Event(%s): %v", obj, event.Sender, event)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event, ok := <-recWatcher.Events():
+			if !ok { // channel shutdown
+				return nil
+			}
+			if err := event.Error; err != nil {
+				return errwrap.Wrapf(err, "Unknown %s watcher error", obj)
+			}
+			if obj.debug { // don't access event.Body if event.Error isn't nil
+				log.Printf("%s: Event(%s): %v", obj, event.Body.Name, event.Body.Op)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+
+		case event := <-svcChan:
+			if obj.debug {
+				log.Printf("%s: Event(%s): %v", obj, event[NetworkdService], event)
+			}
+
+			send = true
+			obj.StateOK(false) // dirty
+		case err := <-svcErr:
+			return errwrap.Wrapf(err, "unknown %s error", obj)
+
+		case event := <-obj.Events():
+			if exit, send = obj.ReadEvent(event); exit != nil {
+				return *exit // exit
+			}
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			send = false
+			obj.Event()
+		}
+	}
+}
+
+// ifaceCheckApply checks the state of the network device and performs the
+// appropriate actions to converge its state with the graph.
+func (obj *NetRes) ifaceCheckApply(apply bool) (checkOK bool, _ error) {
+	// check the interface state.
+	iface, err := net.InterfaceByName(obj.GetName())
+	if err != nil {
+		return false, errwrap.Wrapf(err, "error finding interface: %s", obj.GetName())
+	}
+
+	// check if the interface is up
+	ifaceUp := strings.Contains(iface.Flags.String(), "up")
+
+	// if the interface's state is correct, we're done
+	if obj.State == "down" && !ifaceUp {
+		return true, nil
+	}
+	if obj.State == "up" && ifaceUp {
+		// error if we lose the connection and softfail is false
+		if !obj.SoftFail && obj.oState == dbus.MakeVariant("no-carrier") {
```

Comment:
This is a bug. If we haven't yet received an event from the NIC, obj.oState will be nil. I'll figure out an alternative.